src/listener.js

```diff
@@ -1,183 +1,266 @@
+  Discovery,
+  DiscoveryAck,
+  DialRequest,
+  DialResponse,
+  ErrorTranslations
+} = require('./proto')
 
-const debug = require('debug')
-const log = debug('libp2p:stardust:client')
-
-const EventEmitter = require('events').EventEmitter
+const sha5 = (data) => crypto.createHash('sha512').update(data).digest()
 
 function translateAndThrow (eCode) {
   throw new Error(ErrorTranslations[eCode] || ('Unknown error #' + eCode + '! Please upgrade libp2p-stardust to the latest version!'))
 }
 
-function noop () { }
-
-const ACK = Buffer.from('01', 'hex')
-
+/**
+* Stardust Transport Listener
+* @class
+*/
 class Listener extends EventEmitter {
-  constructor (client, handler) {
+  /**
+   * @constructor
+   * @param {Object} properties - properties for the listener
+   * @param {Stardust} properties.client - Stardust client reference
+   * @param {Upgrader} properties.upgrader - Connection upgrader
+   * @param {function (Connection)} [properties.handler] - New connection handler
+   * @param {Object} [properties.options] - options for the listener
+   * @param {number} [propertis.options.discoveryInterval=10000]
+   */
+  constructor ({ client, upgrader, handler, options = {} }) {
     super()
     this.client = client
     this.handler = handler
+    this.upgrader = upgrader
+    this.options = options
+
+    this.address = undefined
+    this.isConnected = undefined
+    this.aid = undefined
+    this.serverConnection = undefined
+    this.wrappedStream = undefined
   }
 
-  listen (ma, callback) {
-    if (!callback) {
-      callback = noop
-    }
-
-    this._listen(ma).then(() => {
+  /**
+   * Listen on a multiaddr (from stardust server)
+   * @param {Multiaddr} ma multiaddr to listen on
+   * @returns {Promise<void>}
+   */
+  async listen (ma) {
+    try {
+      await this.connectServer(ma)
       this.emit('listening')
-      callback()
-    }, err => {
+    } catch (err) {
       if (this.client.softFail) {
-        this.emit('listening')
-        callback()
         // dials will fail, but that's just about it
-      } else {
-        this.emit('error', err)
-        callback(err)
+        return
       }
-    })
+      this.emit('error', err)
+    }
   }
 
-  getAddrs (cb) {
-    cb(null, this.address ? [this.address] : [])
+  /**
+   * Get listener addresses
+   * @returns {Array<Multiaddr>}
+   */
+  getAddrs () {
+    return this.address ? [this.address] : []
   }
 
-  close () {
-    if (!this.connected) { return }
-    this.connected = false // will prevent new conns, but will keep current ones as interface requires it
-    delete this.client.connections[this.aid]
+  /**
+   * Close listener
+   * @returns {Promise<void>}
+   */
+  async close () {
+    if (!this.isConnected) return
+    this.isConnected = false // will prevent new conns, but will keep current ones as interface requires it
+
+    // close stream and connection with the server
+    const wrappedStream = this.wrappedStream.unwrap()
+    wrappedStream.sink([])
```

Comment:
This should be unnecessary. The call below to `this.serverConnection.close()` is terminating the entire connection, so `this.wrappedStream`, which is a stream of that connection should automatically terminate.

Also, just a note, `wrappedStream.sink([])` **ONLY** closes the write side of the stream, the read would still be open. `source` and `sink` must both be closed. We should document this better in the libp2p and connection interface docs, with examples of how to properly close streams to help people avoid halfopen streams.

Reply:
> We should document this better in the libp2p and connection interface docs, with examples of how to properly close streams to help people avoid halfopen streams.

Yeah, it makes sense